In [15]:
#Importing Libraries
import pandas as pd
from sklearn import metrics
from sklearn.model_selection import train_test_split
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.tree import DecisionTreeClassifier
from imblearn.combine import SMOTEENN

In [16]:
df=pd.read_csv('/content/tel_churn.csv')

In [17]:
df

Unnamed: 0  age  dependents  monthly_charges  total_charges  \
0             0   60           1            67.82         304.67   
1             1   56           0            92.70         168.76   
2             2   24           3            90.22          96.61   
3             3   63           1            70.24         729.03   
4             4   66           1            21.42         813.48   
..          ...  ...         ...              ...            ...   
495         495   35           1            69.94         838.09   
496         496   30           0            26.57         206.29   
497         497   62           0            17.19         523.34   
498         498   34           0            18.91         595.48   
499         499   25           1            92.11         388.23   

     call_duration_minutes  churn_status  gender_Female  gender_Male  \
0                      586             0              1            0   
1                       75             0              1            0   
2                      931             0              0            1   
3                      645             0              1            0   
4                      509             0              0            1   
..                     ...           ...            ...          ...   
495                    110             0              1            0   
496                    281             0              1            0   
497                    121             1              0            1   
498                    492             0              0            1   
499                    213             0              0            1   

     marital_status_Married  ...  online_backup_No  online_backup_Yes  \
0                         0  ...                 0                  1   
1                         0  ...                 0                  1   
2                         1  ...                 0                  1   
3                         1  ...                 0                  1   
4                         0  ...                 1                  0   
..                      ...  ...               ...                ...   
495                       1  ...                 1                  0   
496                       1  ...                 0                  1   
497                       0  ...                 1                  0   
498                       0  ...                 0                  1   
499                       1  ...                 0                  1   

     device_protection_No  device_protection_Yes  tech_support_No  \
0                       0                      1                0   
1                       1                      0                0   
2                       1                      0                0   
3                       1                      0                1   
4                       1                      0                1   
..                    ...                    ...              ...   
495                     1                      0                1   
496                     1                      0                0   
497                     0                      1                1   
498                     0                      1                0   
499                     1                      0                1   

     tech_support_Yes  streaming_tv_No  streaming_tv_Yes  streaming_movies_No  \
0                   1                1                 0                    0   
1                   1                0                 1                    1   
2                   1                1                 0                    0   
3                   0                0                 1                    1   
4                   0                0                 1                    0   
..                ...              ...               ...                  ...   
495                 0                1            

In [18]:
df=df.drop('Unnamed: 0',axis=1)

In [19]:
x=df.drop('churn_status',axis=1)
x

age  dependents  monthly_charges  total_charges  call_duration_minutes  \
0     60           1            67.82         304.67                    586   
1     56           0            92.70         168.76                     75   
2     24           3            90.22          96.61                    931   
3     63           1            70.24         729.03                    645   
4     66           1            21.42         813.48                    509   
..   ...         ...              ...            ...                    ...   
495   35           1            69.94         838.09                    110   
496   30           0            26.57         206.29                    281   
497   62           0            17.19         523.34                    121   
498   34           0            18.91         595.48                    492   
499   25           1            92.11         388.23                    213   

     gender_Female  gender_Male  marital_status_Married  \
0                1            0                       0   
1                1            0                       0   
2                0            1                       1   
3                1            0                       1   
4                0            1                       0   
..             ...          ...                     ...   
495              1            0                       1   
496              1            0                       1   
497              0            1                       0   
498              0            1                       0   
499              0            1                       1   

     marital_status_Single  contract_type_Monthly  ...  online_backup_No  \
0                        1                      0  ...                 0   
1                        1                      1  ...                 0   
2                        0                      0  ...                 0   
3                        0                      1  ...                 0   
4                        1                      0  ...                 1   
..                     ...                    ...  ...               ...   
495                      0                      1  ...                 1   
496                      0                      0  ...                 0   
497                      1                      0  ...                 1   
498                      1                      0  ...                 0   
499                      0                      1  ...                 0   

     online_backup_Yes  device_protection_No  device_protection_Yes  \
0                    1                     0                      1   
1                    1                     1                      0   
2                    1                     1                      0   
3                    1                     1                      0   
4                    0                     1                      0   
..                 ...                   ...                    ...   
495                  0                     1                      0   
496                  1                     1                      0   
497                  0                     0                      1   
498                  1                     0                      1   
499                  1                     1                      0   

     tech_support_No  tech_support_Yes  streaming_tv_No  streaming_tv_Yes  \
0                  0                 1                1                 0   
1                  0                 1                0                 1   
2                  0                 1                1                 0   
3                  1                 0                0                 1   
4                  1                 0                0                 1   
..               ...               ...              ...               ...   
495                1                 0    

In [20]:
y=df['churn_status']
y

0      0
1      0
2      0
3      0
4      0
      ..
495    0
496    0
497    1
498    0
499    0
Name: churn_status, Length: 500, dtype: int64

In [21]:
df.dtypes

age                               int64
dependents                        int64
monthly_charges                 float64
total_charges                   float64
call_duration_minutes             int64
churn_status                      int64
gender_Female                     int64
gender_Male                       int64
marital_status_Married            int64
marital_status_Single             int64
contract_type_Monthly             int64
contract_type_Yearly              int64
internet_service_DSL              int64
internet_service_Fiber_Optic      int64
phone_service_No                  int64
phone_service_Yes                 int64
multiple_lines_No                 int64
multiple_lines_Yes                int64
online_security_No                int64
online_security_Yes               int64
online_backup_No                  int64
online_backup_Yes                 int64
device_protection_No              int64
device_protection_Yes             int64
tech_support_No                   int64


**Train Test Split**

In [22]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2)


**Decision Tree Classifier**

In [23]:
model_dt=DecisionTreeClassifier(criterion = "gini",random_state = 100,max_depth=6, min_samples_leaf=8)


In [24]:
model_dt.fit(x_train,y_train)


DecisionTreeClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

In [25]:
y_pred=model_dt.predict(x_test)
y_pred

array([1, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [79]:
print(model_dt.score(x_test,y_test))

0.76


In [27]:
print(classification_report(y_test, y_pred, labels=[0,1]))

              precision    recall  f1-score   support

           0       0.82      0.90      0.86        80
           1       0.33      0.20      0.25        20

    accuracy                           0.76       100
   macro avg       0.58      0.55      0.55       100
weighted avg       0.72      0.76      0.74       100



As you can see that the accuracy is quite low, and as it's an imbalanced dataset(Ratio:78:22), we shouldn't consider Accuracy as our metrics to measure the model, as Accuracy is cursed in imbalanced datasets.
Hence, we need to check recall, precision & f1 score for the minority class, and it's quite evident that the precision, recall & f1 score is too low for Class 1, i.e. churned customers.
Hence, moving ahead to call SMOTEENN (UpSampling + ENN)

In [28]:
from imblearn.combine import SMOTEENN
sm = SMOTEENN()
X_resampled, y_resampled = sm.fit_resample(x,y)

In [29]:
xr_train,xr_test,yr_train,yr_test=train_test_split(X_resampled, y_resampled,test_size=0.2)


In [30]:
model_dt_smote=DecisionTreeClassifier(criterion = "gini",random_state = 100,max_depth=6, min_samples_leaf=8)


In [31]:
model_dt_smote.fit(xr_train,yr_train)
yr_predict = model_dt_smote.predict(xr_test)
model_score_r = model_dt_smote.score(xr_test, yr_test)
print(model_score_r)
print(metrics.classification_report(yr_test, yr_predict))


0.7471264367816092
              precision    recall  f1-score   support

           0       0.71      0.71      0.71        38
           1       0.78      0.78      0.78        49

    accuracy                           0.75        87
   macro avg       0.74      0.74      0.74        87
weighted avg       0.75      0.75      0.75        87



In [32]:
print(metrics.confusion_matrix(yr_test, yr_predict))


[[27 11]
 [11 38]]


Now we can see quite better results, i.e. Accuracy: 84 %, and a very good recall, precision & f1 score for minority class.

**Random Forest Classifier**

In [33]:
from sklearn.ensemble import RandomForestClassifier

In [34]:
model_rf=RandomForestClassifier(n_estimators=100, criterion='gini', random_state = 100,max_depth=6, min_samples_leaf=8)


In [35]:
model_rf.fit(x_train,y_train)


RandomForestClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

In [36]:
y_pred=model_rf.predict(x_test)


In [82]:
print(model_rf.score(x_test,y_test))


0.8


In [38]:
print(classification_report(y_test, y_pred, labels=[0,1]))


              precision    recall  f1-score   support

           0       0.80      1.00      0.89        80
           1       0.00      0.00      0.00        20

    accuracy                           0.80       100
   macro avg       0.40      0.50      0.44       100
weighted avg       0.64      0.80      0.71       100



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [39]:
sm = SMOTEENN()
X_resampled1, y_resampled1 = sm.fit_resample(x,y)
X_resampled1

age  dependents  monthly_charges  total_charges  call_duration_minutes  \
0     56           0        92.700000     168.760000                     75   
1     24           3        90.220000      96.610000                    931   
2     60           0        76.750000      75.830000                     91   
3     52           2        11.220000     172.690000                    632   
4     24           3        34.440000     689.570000                    497   
..   ...         ...              ...            ...                    ...   
417   26           2        33.602287      61.053973                    355   
418   29           2         9.071470      38.741464                   1105   
419   53           1        75.988725     345.187452                    535   
420   74           1        29.557509     364.024619                   1028   
421   65           0        26.933072     673.545595                    755   

     gender_Female  gender_Male  marital_status_Married  \
0                1            0                       0   
1                0            1                       1   
2                0            1                       1   
3                0            1                       0   
4                1            0                       1   
..             ...          ...                     ...   
417              0            1                       0   
418              0            0                       0   
419              1            0                       0   
420              0            0                       0   
421              0            0                       0   

     marital_status_Single  contract_type_Monthly  ...  online_backup_No  \
0                        1                      1  ...                 0   
1                        0                      0  ...                 0   
2                        0                      0  ...                 1   
3                        1                      0  ...                 0   
4                        0                      0  ...                 1   
..                     ...                    ...  ...               ...   
417                      1                      1  ...                 0   
418                      1                      0  ...                 0   
419                      0                      1  ...                 1   
420                      0                      0  ...                 1   
421                      0                      0  ...                 0   

     online_backup_Yes  device_protection_No  device_protection_Yes  \
0                    1                     1                      0   
1                    1                     1                      0   
2                    0                     0                      1   
3                    1                     1                      0   
4                    0                     0                      1   
..                 ...                   ...                    ...   
417                  1                     0                      1   
418                  1                     1                      0   
419                  0                     0                      0   
420                  0                     0                      1   
421                  0                     0                      0   

     tech_support_No  tech_support_Yes  streaming_tv_No  streaming_tv_Yes  \
0                  0                 1                0                 1   
1                  0                 1                1                 0   
2                  0                 1                1                 0   
3                  1                 0                1                 0   
4                  1                 0                0                 1   
..               ...               ...              ...               ...   
417                0                 1    

In [40]:
xr_train1,xr_test1,yr_train1,yr_test1=train_test_split(X_resampled1, y_resampled1,test_size=0.2)


In [41]:
model_rf_smote=RandomForestClassifier(n_estimators=100, criterion='gini', random_state = 100,max_depth=6, min_samples_leaf=8)


In [42]:
model_rf_smote.fit(xr_train1,yr_train1)


RandomForestClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

In [43]:
yr_predict1 = model_rf_smote.predict(xr_test1)


In [44]:
model_score_r1 = model_rf_smote.score(xr_test1, yr_test1)



In [45]:
print(model_score_r1)
print(metrics.classification_report(yr_test1, yr_predict1))

0.8823529411764706
              precision    recall  f1-score   support

           0       0.87      0.87      0.87        38
           1       0.89      0.89      0.89        47

    accuracy                           0.88        85
   macro avg       0.88      0.88      0.88        85
weighted avg       0.88      0.88      0.88        85



In [46]:
print(metrics.confusion_matrix(yr_test1, yr_predict1))


[[33  5]
 [ 5 42]]


With RF Classifier, also we are able to get better results than Decision Tree.
We can now further go ahead and create multiple classifiers (PCA Model) to see how the model performance is, to compare with RF and DTC.

**Performing PCA**

In [47]:
#Principal Component Analysis
#Dimensionally reduction using singular value decomposition of the data to project in lower dimension.
# Applying PCA

from sklearn.decomposition import PCA
pca = PCA(0.9)
xr_train_pca = pca.fit_transform(xr_train1)
xr_test_pca = pca.transform(xr_test1)
explained_variance = pca.explained_variance_ratio_


In [48]:
model=RandomForestClassifier(n_estimators=100, criterion='gini', random_state = 100,max_depth=6, min_samples_leaf=8)


In [49]:
model.fit(xr_train_pca,yr_train1)


RandomForestClassifier(max_depth=6, min_samples_leaf=8, random_state=100)

In [50]:
yr_predict_pca = model.predict(xr_test_pca)


In [51]:
model_score_r_pca = model.score(xr_test_pca, yr_test1)


In [52]:
print(model_score_r_pca)
print(metrics.classification_report(yr_test1, yr_predict_pca))

0.7294117647058823
              precision    recall  f1-score   support

           0       0.78      0.55      0.65        38
           1       0.71      0.87      0.78        47

    accuracy                           0.73        85
   macro avg       0.74      0.71      0.71        85
weighted avg       0.74      0.73      0.72        85



With PCA, we couldn't see any better results, hence the model which was created by RF Classifier, was the best when compared to DTC and PCA.

###LOGISTIC REGRESSION FOR RESAMPLED DATA

In [59]:
from sklearn.linear_model import LogisticRegression


In [62]:
model_LR = LogisticRegression(solver='liblinear', random_state=1)
model_LR.fit(xr_train, yr_train)


LogisticRegression(random_state=1, solver='liblinear')

In [63]:
yLR_predict = model_LR.predict(xr_test1)

In [64]:
yLR_predict

array([0, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 0,
       1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 1, 1, 0,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1,
       0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 0, 1, 1])

In [65]:
model_score_r_LR = model.score(xr_test1, yr_test1)

In [66]:
print(model_score_r_LR)
print(metrics.classification_report(yr_test1, yLR_predict))

0.9176470588235294
              precision    recall  f1-score   support

           0       0.88      0.95      0.91        38
           1       0.95      0.89      0.92        47

    accuracy                           0.92        85
   macro avg       0.92      0.92      0.92        85
weighted avg       0.92      0.92      0.92        85



Logistic model done using SMOTEENN re-sampled data, outperformed RF classifier when compared with precision and f-1 score, whereas accuracy remains same.

The finalised model is Logistic Regression with SMOTEENN(re-sampled data), is now ready and dumped as model.sav.

**Making a Predictive System**

A quick glance of how the model works!!

In [67]:
import numpy as np

In [68]:
#input_data = ('age,dependents,monthly_charges,total_charges,call_duration_minutes,gender_Female,gender_Male,marital_status_Married,marital_status_Single,contract_type_Monthly,contract_type_Yearly,internet_service_DSL,internet_service_Fiber Optic,phone_service_No,phone_service_Yes,multiple_lines_No,multiple_lines_Yes,online_security_No,online_security_Yes,online_backup_No,online_backup_Yes,device_protection_No,device_protection_Yes,tech_support_No,tech_support_Yes,streaming_tv_No,streaming_tv_Yes,streaming_movies_No,streaming_movies_Yes')
input_data =(60,1,67.82,304.67,586,1,0,0,1,0,1,0,1,0,1,0,1,1,0,0,1,0,1,0,1,1,0,0,1)
#changing the input to numpy array
input_data_as_numpy_array = np.asarray(input_data)

#reshaping the array as we are predicting for one instance
input_data_reshaped = input_data_as_numpy_array.reshape(1,-1)

prediction = model_LR.predict(input_data_reshaped)
print(prediction)

if (prediction[0] == 0):
  print('The person is likely to be CONTINUED!')
else:
  print('The person is likely to be CHURNED!')


[0]
The person is likely to be CONTINUED!


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(


**Pickling the model**










Saving the model for future deployment.

In [69]:
import pickle


In [70]:
filename = 'model.sav'


In [71]:
pickle.dump(model_LR, open(filename, 'wb'))


In [72]:
#loaded the saved model
loaded_model = pickle.load(open('model.sav', 'rb'))


In [73]:
model_score_LR = loaded_model.score(xr_test1, yr_test1)
model_score_LR


0.9176470588235294

The Logistic Regression + SMOTEENN model displays 91.76% of accuracy score. which is far better than DTC, RF, and PCA altogether.

###Quick glance

In [80]:
#DTC MODEL WITH IMBALANCED DATA

print(model_dt.score(x_test,y_test))
print(classification_report(y_test, y_pred, labels=[0,1]))

0.76
              precision    recall  f1-score   support

           0       0.80      1.00      0.89        80
           1       0.00      0.00      0.00        20

    accuracy                           0.80       100
   macro avg       0.40      0.50      0.44       100
weighted avg       0.64      0.80      0.71       100



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [81]:
#DTC MODEL WITH RE-SAMPLED DATA

print(model_score_r)
print(metrics.classification_report(yr_test, yr_predict))

0.7471264367816092
              precision    recall  f1-score   support

           0       0.71      0.71      0.71        38
           1       0.78      0.78      0.78        49

    accuracy                           0.75        87
   macro avg       0.74      0.74      0.74        87
weighted avg       0.75      0.75      0.75        87



In [83]:
#RF MODEL WITH IMBALANCED DATA

print(model_rf.score(x_test,y_test))
print(classification_report(y_test, y_pred, labels=[0,1]))

0.8
              precision    recall  f1-score   support

           0       0.80      1.00      0.89        80
           1       0.00      0.00      0.00        20

    accuracy                           0.80       100
   macro avg       0.40      0.50      0.44       100
weighted avg       0.64      0.80      0.71       100



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


In [84]:
#RF MODEL WITH RE-SAMPLED DATA

print(model_score_r1)
print(metrics.classification_report(yr_test1, yr_predict1))

0.8823529411764706
              precision    recall  f1-score   support

           0       0.87      0.87      0.87        38
           1       0.89      0.89      0.89        47

    accuracy                           0.88        85
   macro avg       0.88      0.88      0.88        85
weighted avg       0.88      0.88      0.88        85



In [85]:
#PCA MODEL WITH RE-SAMPLED DATA

print(model_score_r_pca)
print(metrics.classification_report(yr_test1, yr_predict_pca))

0.7294117647058823
              precision    recall  f1-score   support

           0       0.78      0.55      0.65        38
           1       0.71      0.87      0.78        47

    accuracy                           0.73        85
   macro avg       0.74      0.71      0.71        85
weighted avg       0.74      0.73      0.72        85



In [86]:
#LR MODEL WITH RE-SAMPLED DATA

print(model_score_r_LR)
print(metrics.classification_report(yr_test1, yLR_predict))

0.9176470588235294
              precision    recall  f1-score   support

           0       0.88      0.95      0.91        38
           1       0.95      0.89      0.92        47

    accuracy                           0.92        85
   macro avg       0.92      0.92      0.92        85
weighted avg       0.92      0.92      0.92        85

